File: scripts/eval-expression-suite.mjs
import fs from "fs";
import path from "path";
import { spawnSync } from "child_process";

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (!item.startsWith("--")) continue;
    const key = item.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      out[key] = true;
      continue;
    }
    out[key] = next;
    i += 1;
  }
  return out;
}

function runEval({ inputPath, outputPath, style }) {
  const args = [
    "scripts/eval-auto-expression.mjs",
    "--input",
    inputPath,
    "--output",
    outputPath
  ];
  if (style) args.push("--style", style);

  const res = spawnSync(process.execPath, args, {
    cwd: process.cwd(),
    stdio: "pipe",
    encoding: "utf8"
  });
  if (res.status !== 0) {
    throw new Error(`eval_failed file=${inputPath}\n${res.stderr || res.stdout}`);
  }
}

function safeAvg(items) {
  if (!items.length) return 0;
  return items.reduce((s, n) => s + n, 0) / items.length;
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const testsDir = path.resolve(String(args.dir || "tests/expressions"));
  const reportDir = path.resolve(String(args.outdir || "outputs/eval_suite"));
  const style = String(args.style || "tegang").trim();
  const maxAutoDelta = Number(args["max-auto-delta"] ?? 9.8);
  const maxOverrideDelta = Number(args["max-override-delta"] ?? 10.2);
  const minAvgOverrideSegments = Number(args["min-avg-override-segments"] ?? 0.8);
  const maxAvgChangedSegments = Number(args["max-avg-changed-segments"] ?? 1.6);

  if (!fs.existsSync(testsDir)) {
    throw new Error(`Tests directory not found: ${testsDir}`);
  }
  if (!fs.existsSync(reportDir)) fs.mkdirSync(reportDir, { recursive: true });

  const files = fs
    .readdirSync(testsDir, { withFileTypes: true })
    .filter((d) => d.isFile() && d.name.toLowerCase().endsWith(".txt"))
    .map((d) => d.name)
    .sort();

  if (!files.length) {
    throw new Error(`No .txt test files found in ${testsDir}`);
  }

  const rows = [];
  for (const file of files) {
    const inputPath = path.join(testsDir, file);
    const outPath = path.join(reportDir, `${file.replace(/\.txt$/i, "")}.json`);
    runEval({ inputPath, outputPath: outPath, style });
    const report = JSON.parse(fs.readFileSync(outPath, "utf8"));
    rows.push({
      file,
      segments: Number(report?.segments || 0),
      changedIntentSegments: Number(report?.metrics?.changedIntentSegments || 0),
      overrideIntentSegments: Number(report?.metrics?.overrideIntentSegments || 0),
      autoTransitionDelta: Number(report?.metrics?.autoTransitionDelta || 0),
      overrideTransitionDelta: Number(report?.metrics?.overrideTransitionDelta || 0)
    });
  }

  const summary = {
    at: new Date().toISOString(),
    dir: path.relative(process.cwd(), testsDir),
    style,
    caseCount: rows.length,
    metrics: {
      avgSegments: Number(safeAvg(rows.map((r) => r.segments)).toFixed(3)),
      avgChangedIntentSegments: Number(safeAvg(rows.map((r) => r.changedIntentSegments)).toFixed(3)),
      avgOverrideIntentSegments: Number(safeAvg(rows.map((r) => r.overrideIntentSegments)).toFixed(3)),
      avgAutoTransitionDelta: Number(safeAvg(rows.map((r) => r.autoTransitionDelta)).toFixed(3)),
      avgOverrideTransitionDelta: Number(safeAvg(rows.map((r) => r.overrideTransitionDelta)).toFixed(3))
    }
  };

  const gateChecks = {
    autoTransitionOk: summary.metrics.avgAutoTransitionDelta <= maxAutoDelta,
    overrideTransitionOk: summary.metrics.avgOverrideTransitionDelta <= maxOverrideDelta,
    overrideCoverageOk: summary.metrics.avgOverrideIntentSegments >= minAvgOverrideSegments,
    changedIntentOk: summary.metrics.avgChangedIntentSegments <= maxAvgChangedSegments
  };
  const passed = Object.values(gateChecks).every(Boolean);
  summary.gate = {
    status: passed ? "pass" : "fail",
    thresholds: {
      maxAutoDelta,
      maxOverrideDelta,
      minAvgOverrideSegments,
      maxAvgChangedSegments
    },
    checks: gateChecks
  };
  summary.cases = rows.map((row) => ({
    ...row,
    gate: {
      autoDeltaOk: row.autoTransitionDelta <= maxAutoDelta,
      overrideDeltaOk: row.overrideTransitionDelta <= maxOverrideDelta,
      overrideCoverageOk: row.overrideIntentSegments >= 1
    }
  }));

  const summaryJsonPath = path.join(reportDir, "summary.json");
  fs.writeFileSync(summaryJsonPath, JSON.stringify(summary, null, 2), "utf8");

  const mdLines = [
    "# Expression Eval Suite Summary",
    "",
    `- Generated at: ${summary.at}`,
    `- Test dir: \`${summary.dir}\``,
    `- Style: \`${summary.style}\``,
    `- Cases: ${summary.caseCount}`,
    "",
    "## Aggregate Metrics",
    "",
    `- avgSegments: ${summary.metrics.avgSegments}`,
    `- avgChangedIntentSegments: ${summary.metrics.avgChangedIntentSegments}`,
    `- avgOverrideIntentSegments: ${summary.metrics.avgOverrideIntentSegments}`,
    `- avgAutoTransitionDelta: ${summary.metrics.avgAutoTransitionDelta}`,
    `- avgOverrideTransitionDelta: ${summary.metrics.avgOverrideTransitionDelta}`,
    "",
    "## Quality Gate",
    "",
    `- status: **${summary.gate.status.toUpperCase()}**`,
    `- autoTransitionOk: ${summary.gate.checks.autoTransitionOk}`,
    `- overrideTransitionOk: ${summary.gate.checks.overrideTransitionOk}`,
    `- overrideCoverageOk: ${summary.gate.checks.overrideCoverageOk}`,
    `- changedIntentOk: ${summary.gate.checks.changedIntentOk}`,
    "",
    "## Cases",
    "",
    "| file | segments | changedIntent | overrideIntent | autoDelta | overrideDelta | gate |",
    "|---|---:|---:|---:|---:|---:|---|"
  ];
  for (const row of summary.cases) {
    const rowPass = row.gate.autoDeltaOk && row.gate.overrideDeltaOk && row.gate.overrideCoverageOk;
    mdLines.push(
      `| ${row.file} | ${row.segments} | ${row.changedIntentSegments} | ${row.overrideIntentSegments} | ${row.autoTransitionDelta} | ${row.overrideTransitionDelta} | ${rowPass ? "pass" : "fail"} |`
    );
  }
  const summaryMdPath = path.join(reportDir, "summary.md");
  fs.writeFileSync(summaryMdPath, `${mdLines.join("\n")}\n`, "utf8");

  console.log(
    `eval_suite_done cases=${rows.length} gate=${summary.gate.status} summary_json=${path.relative(process.cwd(), summaryJsonPath)} summary_md=${path.relative(process.cwd(), summaryMdPath)}`
  );

  if (!passed) {
    process.exitCode = 2;
  }
}

main();

File: scripts_py/eval_expression_suite.py
import json
import os
from pathlib import Path
from datetime import datetime, timezone

from common_pipeline import parse_args
from eval_auto_expression import evaluate_auto_expression


def safe_avg(items):
    return (sum(items) / len(items)) if items else 0.0


def main():
    args = parse_args(os.sys.argv[1:])
    tests_dir = Path(str(args.get("dir") or "tests/expressions")).resolve()
    report_dir = Path(str(args.get("outdir") or "outputs/eval_suite")).resolve()
    style = str(args.get("style") or "tegang").strip()
    profile_file = str(args.get("profile-file") or "").strip() or None
    hybrid_prosody = str(args.get("hybrid-prosody", "true")).lower() == "true"
    humanize_intensity = float(args.get("humanize-intensity") or 0.7)
    prosody_limiter = str(args.get("prosody-limiter", "true")).lower() == "true"
    prosody_limiter_strength = float(args.get("prosody-limiter-strength") or 0.64)
    auto_expressive = str(args.get("auto-expressive", "true")).lower() == "true"
    max_auto_delta = float(args.get("max-auto-delta") or 9.8)
    max_override_delta = float(args.get("max-override-delta") or 10.2)
    min_avg_override_segments = float(args.get("min-avg-override-segments") or 0.8)
    max_avg_changed_segments = float(args.get("max-avg-changed-segments") or 1.6)

    if not tests_dir.exists():
        raise RuntimeError(f"Tests directory not found: {tests_dir}")
    report_dir.mkdir(parents=True, exist_ok=True)
    files = sorted([p for p in tests_dir.iterdir() if p.is_file() and p.suffix.lower() == ".txt"])
    if not files:
        raise RuntimeError(f"No .txt test files found in {tests_dir}")

    rows = []
    for fp in files:
        out_path = report_dir / f"{fp.stem}.json"
        report = evaluate_auto_expression(
            input_path=str(fp),
            style=style,
            profile_file=profile_file,
            humanize_intensity=humanize_intensity,
            hybrid_prosody=hybrid_prosody,
            prosody_limiter=prosody_limiter,
            prosody_limiter_strength=prosody_limiter_strength,
            auto_expressive=auto_expressive,
        )
        with open(out_path, "w", encoding="utf-8") as f:
            json.dump(report, f, indent=2, ensure_ascii=False)
        m = report.get("metrics", {})
        rows.append(
            {
                "file": fp.name,
                "segments": int(report.get("segments", 0)),
                "changedIntentSegments": float(m.get("changedIntentSegments", 0)),
                "overrideIntentSegments": float(m.get("overrideIntentSegments", 0)),
                "autoTransitionDelta": float(m.get("autoTransitionDelta", 0)),
                "overrideTransitionDelta": float(m.get("overrideTransitionDelta", 0)),
                "autoProsodyEnergy": float(m.get("autoProsodyEnergy", 0)),
                "overrideProsodyEnergy": float(m.get("overrideProsodyEnergy", 0)),
            }
        )

    summary = {
        "at": datetime.now(timezone.utc).isoformat(),
        "dir": str(tests_dir.relative_to(Path.cwd())).replace("\\", "/"),
        "style": style,
        "profileFile": profile_file or "active",
        "hybridProsody": hybrid_prosody,
        "caseCount": len(rows),
        "metrics": {
            "avgSegments": round(safe_avg([r["segments"] for r in rows]), 3),
            "avgChangedIntentSegments": round(safe_avg([r["changedIntentSegments"] for r in rows]), 3),
            "avgOverrideIntentSegments": round(safe_avg([r["overrideIntentSegments"] for r in rows]), 3),
            "avgAutoTransitionDelta": round(safe_avg([r["autoTransitionDelta"] for r in rows]), 3),
            "avgOverrideTransitionDelta": round(safe_avg([r["overrideTransitionDelta"] for r in rows]), 3),
            "avgAutoProsodyEnergy": round(safe_avg([r["autoProsodyEnergy"] for r in rows]), 3),
            "avgOverrideProsodyEnergy": round(safe_avg([r["overrideProsodyEnergy"] for r in rows]), 3),
        },
    }
    checks = {
        "autoTransitionOk": summary["metrics"]["avgAutoTransitionDelta"] <= max_auto_delta,
        "overrideTransitionOk": summary["metrics"]["avgOverrideTransitionDelta"] <= max_override_delta,
        "overrideCoverageOk": summary["metrics"]["avgOverrideIntentSegments"] >= min_avg_override_segments,
        "changedIntentOk": summary["metrics"]["avgChangedIntentSegments"] <= max_avg_changed_segments,
    }
    passed = all(bool(v) for v in checks.values())
    summary["gate"] = {
        "status": "pass" if passed else "fail",
        "thresholds": {
            "maxAutoDelta": max_auto_delta,
            "maxOverrideDelta": max_override_delta,
            "minAvgOverrideSegments": min_avg_override_segments,
            "maxAvgChangedSegments": max_avg_changed_segments,
        },
        "checks": checks,
    }
    summary["cases"] = []
    for row in rows:
        summary["cases"].append(
            {
                **row,
                "gate": {
                    "autoDeltaOk": row["autoTransitionDelta"] <= max_auto_delta,
                    "overrideDeltaOk": row["overrideTransitionDelta"] <= max_override_delta,
                    "overrideCoverageOk": row["overrideIntentSegments"] >= 1,
                },
            }
        )

    summary_json = report_dir / "summary.json"
    with open(summary_json, "w", encoding="utf-8") as f:
        json.dump(summary, f, indent=2, ensure_ascii=False)

    md_lines = [
        "# Expression Eval Suite Summary",
        "",
        f"- Generated at: {summary['at']}",
        f"- Test dir: `{summary['dir']}`",
        f"- Style: `{summary['style']}`",
        f"- Cases: {summary['caseCount']}",
        "",
        "## Aggregate Metrics",
        "",
        f"- avgSegments: {summary['metrics']['avgSegments']}",
        f"- avgChangedIntentSegments: {summary['metrics']['avgChangedIntentSegments']}",
        f"- avgOverrideIntentSegments: {summary['metrics']['avgOverrideIntentSegments']}",
        f"- avgAutoTransitionDelta: {summary['metrics']['avgAutoTransitionDelta']}",
        f"- avgOverrideTransitionDelta: {summary['metrics']['avgOverrideTransitionDelta']}",
        f"- avgAutoProsodyEnergy: {summary['metrics']['avgAutoProsodyEnergy']}",
        f"- avgOverrideProsodyEnergy: {summary['metrics']['avgOverrideProsodyEnergy']}",
        "",
        "## Quality Gate",
        "",
        f"- status: **{summary['gate']['status'].upper()}**",
        f"- autoTransitionOk: {summary['gate']['checks']['autoTransitionOk']}",
        f"- overrideTransitionOk: {summary['gate']['checks']['overrideTransitionOk']}",
        f"- overrideCoverageOk: {summary['gate']['checks']['overrideCoverageOk']}",
        f"- changedIntentOk: {summary['gate']['checks']['changedIntentOk']}",
        "",
        "## Cases",
        "",
        "| file | segments | changedIntent | overrideIntent | autoDelta | overrideDelta | gate |",
        "|---|---:|---:|---:|---:|---:|---|",
    ]
    for row in summary["cases"]:
        row_pass = row["gate"]["autoDeltaOk"] and row["gate"]["overrideDeltaOk"] and row["gate"]["overrideCoverageOk"]
        md_lines.append(
            f"| {row['file']} | {row['segments']} | {row['changedIntentSegments']} | {row['overrideIntentSegments']} | "
            f"{row['autoTransitionDelta']} | {row['overrideTransitionDelta']} | {'pass' if row_pass else 'fail'} |"
        )
    summary_md = report_dir / "summary.md"
    with open(summary_md, "w", encoding="utf-8") as f:
        f.write("\n".join(md_lines) + "\n")

    print(
        f"eval_suite_done cases={len(rows)} gate={summary['gate']['status']} "
        f"summary_json={str(summary_json.relative_to(Path.cwd())).replace('\\', '/')} "
        f"summary_md={str(summary_md.relative_to(Path.cwd())).replace('\\', '/')}"
    )
    if not passed:
        raise SystemExit(2)


if __name__ == "__main__":
    main()
